### DISCUSSION TITLE
`--config` command-line option

### DISCUSSION THREAD

**Discussion by ilyagr** [jj-vcs/jj#688] [Ideas]
What do you think about having a `--config TOML_STRING` global JJ option? This could replace the `--color` option, you would do `--config ui.color=always` instead. You could also do `--config ui.diff-editor=vimdiff`. You could also merge in an extra TOML file with `--config "$(cat extra-config.toml)"` on bash.

The downside of TOML is that it often requires line breaks, which are inconvenient on the command-line. However, that can be ameliorated by allowing repetition of `--config` options.

It could also be shortened to `-c`, assuming that global options are in a separate namespace from the subcommand options.

This came up as I was thinking about whether the `resolve` command discussed in https://github.com/martinvonz/jj/issues/18 needs a `--tool` option.

**Comment by martinvonz** [comment]
Yes, I agree that we should do that. Both Git and Mercurial do it. I just haven't gotten around to it.

The option would go in the [`GlobalArgs` struct](https://github.com/martinvonz/jj/blob/55184237e1b757b79ab81761f540b3d6089fef3f/src/cli_util.rs#L1128).

One thing that complicates it a bit is aliases. I know that Git lets you define an alias that adds another alias, so you can do `git -c alias.foo='-c alias.bar=status bar' foo`. I'm not sure how useful that is, but it seems that we should ideally support it.

Oh, another thing that may complicate it is the planned support for reading configs from `.jj/config.toml` (or something like that). Since we have a `-R` flag for specifying a repo, it seems like there is another level of recursion like for aliases. For example, let's say you're in `~/repo1` and run `jj foo`, and there's a `~/repo/.jj/config.toml` that says `alias.foo = -R ~/repo2 foo`, I suppose that means we should go to that repo and load the repo-level config from there (which could lead to infinite recursion if that repo's config says `alias.foo = -R ~/repo1 foo`). I checked what Git does in this case (the simple case without mutual recursion). It says `fatal: alias 'foo' changes environment variables.`.

I don't know if there are any security issues to think about.

**Reply by ilyagr** [reply]
Crazy idea: what if `jj` *didn't have* general aliases? The shell or `clap` seem better positioned to deal with these issues.

I would very much like aliases inside revsets or for log output formats, but these specific cases don't raise nearly as many issues. Not having general aliases would make it more difficult and more important to carefully design those, as well as configuration for default revset patterns or formats. But, perhaps,  that's worthwhile.
